File: Online-judge/leetcode/missing-element-in-sorted-array.py
from typing import List


class Solution:
    def missingElement(self, nums: List[int], k: int) -> int:
        n = len(nums)

        missingNums = 0
        for i in range(1, n):
            missingNums = nums[i] - nums[i - 1] - 1 # pair missing number count
            if missingNums >= k:
                return nums[i - 1] + k
            k -= missingNums # update found missing number
        return nums[n - 1] + k 

File: Online-judge/leetcode/missing-element-in-sorted-array.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * @author Md.Shohag <shohag.fks@gmail.com>
 * Problem : https://leetcode.com/problems/missing-element-in-sorted-array/
 */

/**
 * @brief 
 * Time complexity : O(logn)
 * Space complexity: O(1)
 * Solution: 
 *  number of element = nums[i] - nums[0] + 1
 *  missing element = (nums[i] - nums[0] + 1) - (i - 1) // (i element existing in array)
 *  = nums[i] - nums[0] - 1 = nums[mid] - nums[0] - mid
 */


class Solution {
public:
    int missingElement(vector<int>& nums, int k) {
        int left = 0, right = nums.size() - 1;

        while (left < right) {
            int mid = right - (right - left) / 2;
            if (nums[mid] - nums[0] - mid < k) {
                left = mid;
            } else {
                right = mid - 1;
            }
        }
        return nums[0] + k + left;
    }
};